Firestore/Example/Tests/SpecTests/FSTMockDatastore.mm

```diff
@@ -162,7 +163,8 @@ void WriteWatchChange(const WatchChange& change, SnapshotVersion snap) {
                   GrpcConnection* grpc_connection,
                   WriteStreamCallback* callback,
                   MockDatastore* datastore)
-      : WriteStream{worker_queue, credentials_provider, serializer, grpc_connection, callback},
+      : WriteStream{worker_queue, std::shared_ptr<CredentialsProvider>(credentials_provider),
```

Comment:
This is very dangerous.

A `shared_ptr` has a constructor that takes a pointer, but it presumes it to be a pointer allocated by `new` and will call `delete` on it once the reference counter reaches zero. In this case, however, the pointer is _not_ allocated by `new` -- this is initialized by [passing the address of a local variable](https://github.com/firebase/firebase-ios-sdk/blob/5cfa39c6205d9001fa920e57dd9fc07e6f61ef43/Firestore/Example/Tests/Integration/FSTDatastoreTests.mm#L188). Trying to delete a stack variable is undefined behavior.

In general, it's best to always use `std::make_shared` in preference to passing a pointer directly to `shared_ptr`'s constructor. This would have prevented this problem.

The easiest way to fix it is probably to remove `_credentials` data member from `FSTDatastoreTests` and just create the shared pointer "on the fly":
```cpp
_datastore = std::make_shared<Datastore>(
      _databaseInfo, _testWorkerQueue, std::make_shared<EmptyCredentialsProvider>());
```

Reply:
Good suggestion!